# Week 1: Python Basics & Flow Control
AI for Good - Minor ADSAI

Work through the sections below in order. Write your code in the empty cells.
Run a cell with Shift + Enter. Getting errors is part of the method, not a
sign that something is wrong.

Today runs in **three learning waves**. Each wave starts with a short
mini-lecture, then you code. A wave starts at every horizontal line in this
notebook, marked like this:

**LEARNING WAVE 1 · SLIDES 5-8**

- Wave 1 - section 1: Python as a calculator
- Wave 2 - section 2: Variables and types
- Wave 3 - section 3: Making decisions
- Then section 4: the big individual assignment, and sections 5-7: wrap-up
  (GitHub portfolio, AI app builders, and what to remember)

Do not run ahead into the next wave - wait for the mini-lecture.

---

**LEARNING WAVE 1 · SLIDES 5-8, 30 minutes**

## 1. Python as a calculator

Sections 1.1 to 1.4. The arithmetic operators, printing results, and your first calculation with real climate data.

### 1.1 Arithmetic symbols

Python uses these symbols to calculate:

- Add: `+`
- Subtract: `-`
- Multiply: `*`
- Divide: `/`
- Integer division: `//`
- Modulo: `%`
- Exponent: `**`

Show the result of:
- 8 times 10
- 10 to the power of 3
- 300 minus 25 times 10 (check whether the normal order of operations is applied)

Use the 3 cells below for your answers.

In [1]:
print(8*10)

80


In [2]:
print(10**3)

1000


In [3]:
print(300 - (25 * 10))

50


### 1.2 Printing results

A cell only shows the result of its *last* line. Use the print function to show
the results of all three calculations below in one cell:
- 8 times 10
- 10 to the power of 3
- 300 minus 25 times 10

In [4]:
print(8*10)
print(10**3)
print(300 - (25 * 10))


80
1000
50


- What happens if you put a hashtag (#) in front of one of the print statements?

Your answer: The line becomes a comment, so Python ignores it completely:
that print statement is not executed and its output disappears from the cell.
The `#` is how you add notes to your code for humans.

### 1.3 Working out the operators

*Slide 7*

Nobody has told you yet what `//`, `%` and `**` do. You are going to work that
out yourself.

- First write down your guess for each of the four lines below.
- Then run the cell and compare the output with your guess.

I think that the answers will be:
3.5
3
4
49

In [5]:
print(7 / 2)
print(7 // 2)
print(7 % 2)
print(7 ** 2)

3.5
3
1
49


- Run the cell below as well, and look at both outputs together.

In [6]:
print(9 / 3)
print(9 // 3)
print(9 % 3)
print(11 % 3)
print(10 % 2)
print(11 % 2)

3.0
3
0
2
0
1


- Explain in your own words what each of these four operators does:
`/`, `//`, `%` and `**`.

Your answer: The / operator divides the numbers. The // divides and rounds the number down to the whole number. The % gives the the remainder of the division. The ** operator gives the calculation of the power operator in mathematics.


- One of these operators answers a question programmers ask constantly:
*is this number even or odd?* Which one, and how? Show it in the cell below
for the numbers 10 and 11.

In [7]:
print(10 % 2)
print(11 % 2)

0
1


### 1.4 Climate arithmetic

*Slide 8 - 15 minutes*

A return flight from Amsterdam to New York emits about 1258 kg of CO2 per
passenger. This is real data - your code is working with real-world impact.

- Print the total CO2 for 3 passengers, using an expression. No variables yet:
just numbers and operators.

In [8]:
print(1258 * 3)

3774


---

**LEARNING WAVE 2 · SLIDES 9-11, 40 minutes**

## 2. Variables and types

Sections 2.1 to 2.6. Storing values under a name, the four data types, converting between them, and asking the user a question.

### 2.1 Variables

- Define the following variables and assign values to them:

`flight_co2`:
the emission per passenger is 1258.

`passengers`:
the number of passengers is 3.

`total_co2`:
the total emission is the emission per passenger times the number of passengers.

- Use the variables to print the total.

In [9]:
flight_co2 = 1258
passengers = 3
total_co2 = flight_co2 * passengers
print(total_co2)

3774


### 2.2 Naming conventions

- What are the most important conventions for variable names in Python?
- Why is it important to stick to them?

Your answer: Use lowercase names with underscores between words (snake_case),
like `flight_co2` or `total_co2`. Make the name describe what the value is.
Constants are written in ALL_CAPS. Never use spaces, never start with a digit,
and avoid Python keywords and built-in function names such as `max`, `print` or
`type`. It is important to stick to these conventions because code is read far
more often than it is written: consistent, descriptive names make your code
understandable for other people and for your future self - and using a built-in
name can even break Python, as the `max` example shows.

- Run the code in the cell below. The result is 1258, because that is the
highest number in the list.
- Now remove the hashtag in front of `max = 15` and run the cell again. What happens?

In [10]:
#max = 15
highest = max([1258, 400, 230])
print(highest)

1258


Your answer: The first run prints 1258, because `max()` is Python's built-in
function that returns the largest number of the list. After removing the
hashtag, `max = 15` turns `max` into an ordinary integer variable, and the line
`max([1258, 400, 230])` now tries to call that integer as if it were a
function. Python gives an error (TypeError: 'int' object is not callable), and
the built-in `max()` stays broken until you restart the kernel.

- You now have to restart the kernel to make `max()` work again.

### 2.3 Types

- Apply the `type()` function to determine the data type of `flight_co2`,
`passengers` and `total_co2`. Print the results.

In [11]:
print(type(flight_co2))
print(type(passengers))
print(type(total_co2))

<class 'int'>
<class 'int'>
<class 'int'>


- What is the difference between an integer and a float?

Your answer: An integer is a whole number without a decimal point, like `3` or
`1258`. A float is a number with a decimal point, like `3.0` or `0.105`. Floats
are used whenever a value can have a fractional part, for example the result of
a division or an energy use of 0.105 kWh.

- Create variables with your name and whether you cycle to university.
For example:

`my_name = "Amara Okonkwo"`
`cycles = True`

- Print the types of both variables.

In [12]:
my_name = "Alex Wrobel"
cycles = True
print(type(my_name))
print(type(cycles))

<class 'str'>
<class 'bool'>


- Explain what the following types are: string, integer, float and Boolean.

Your answer: A string is text, written between quotes, like `"hello"`. An
integer is a whole number, like `3`. A float is a number with a decimal point,
like `3.5`. A Boolean is a value that can only be `True` or `False`; every
comparison gives back a Boolean.

- Sometimes numbers are stored as strings. The variable `number` below
contains the string `'1258'`.
- Convert it to an integer and print the type of the result.

In [13]:
number = '1258'
number = int(number)
print(type(number))

<class 'int'>


### 2.4 Operations with strings

- Convert `number` back to a string.
- Create the variable `add_number` that contains `number` plus `number`.
- Print `add_number`.

In [14]:
number = str(number)
add_number = number + number
print(add_number)

12581258


You can see here that the `+` operator behaves differently for a string
than it does for an integer or a float.

### 2.5 Printing combinations of different data types

To print variables of different types with one print function, you have to
convert them all to strings first. You can then join them with `+`.

In section 2.1 you made the variables `passengers` and `total_co2`. They have
different data types.

- Print the following sentence using both variables:

`A group of 3 passengers emits 3774 kg CO2 on this flight.`

In [15]:
# hint: print('A group of ' + str(passengers) + ______________ )
print('A group of ' + str(passengers) + ' passengers emits ' + str(total_co2) + ' kg CO2 on this flight.')

A group of 3 passengers emits 3774 kg CO2 on this flight.


### 2.6 Talking to the user

*Slide 11 - 15 minutes*

Training a large AI model can emit as much CO2 as five cars over their entire
lifetime, and every chat with an AI model uses energy too. You are going to
estimate your own AI energy footprint.

- Ask the user for their name using `input()`.
- Ask how many AI chats they have per day, and convert the answer to an integer.
- Calculate `weekly_chats`: the daily number times 7.
- One chat uses about 0.003 kWh. Calculate `weekly_energy`.
- Print: `Hi [name]! Your weekly AI energy use is [X] kWh.`

In [16]:
name = input('What is your name? ')
chats_per_day = int(input('How many AI chats do you have per day? '))
weekly_chats = chats_per_day * 7
weekly_energy = weekly_chats * 0.003
print('Hi ' + name + '! Your weekly AI energy use is ' + str(weekly_energy) + ' kWh.')

What is your name? Alex Wrobel


How many AI chats do you have per day? 5


Hi Alex Wrobel! Your weekly AI energy use is 0.105 kWh.


- Which data type does `input()` always return? Why do you have to convert it?

Your answer: `input()` always returns a string, even when the user types a
number - the digits arrive as text. You have to convert the answer with `int()`
or `float()` because Python cannot do arithmetic with strings (and `+` on two
strings joins them into one, it does not add them up).

---

**LEARNING WAVE 3 · SLIDES 12-15, 20 minutes**

## 3. Making decisions

Sections 3.1 to 3.4. Comparisons, Booleans, and the if / elif / else block that is the core of every rule-based decision system.

### 3.1 Comparison operators

Python uses these symbols to compare two values:

- Equal to: `==`
- Not equal to: `!=`
- Greater than: `>`
- Less than: `<`
- Greater than or equal to: `>=`
- Less than or equal to: `<=`

Every comparison gives back a Boolean: `True` or `False`.

- Use print to answer the questions below:
- Is 5 equal to 5?
- Is 5 equal to `'5'`?
- Is the total CO2 of 3 passengers greater than 3000?

In [17]:
print(5 == 5)
print(5 == '5')
print(total_co2 > 3000)

True
False
True


- Explain in your own words the difference between `=` and `==`.

Your answer: One `=` is assignment: it stores a value in a variable, like
`total_co2 = 3774`. Two `==` is comparison: it checks whether two values are
equal and gives back `True` or `False`, like `5 == 5` gives `True`. Using `=`
where you meant `==` is a classic beginner bug.

### 3.2 Boolean operators

`and` gives True only when both sides are True. `or` gives True when at least
one side is True. `not` flips the value.

- Write down your guess for each line, then run the cell.

My guesses: `False`, `True`, `False`, `False` - the last one because `3 > 5`
is False, and `True and False` is False.

In [18]:
print(True and False)
print(True or False)
print(not True)
print(1258 > 1000 and 3 > 5)

False
True
False
False


### 3.3 if / elif / else

*Slide 15 - 20 minutes*

Many AI systems make decisions with exactly the structure you are about to
write. The same logic that flags a "high-risk loan applicant" can reproduce
human bias.

Python checks the conditions from top to bottom and runs the first one that is
True. Indentation is not styling in Python - it is the syntax that decides
which lines belong to which branch.

- Create the variable `income` with `income = int(input('Annual income in euros: '))`
- Write an `if` / `elif` / `else` block that assigns a risk level:
  - `income` below 20000 gives `'high risk'`
  - `income` below 40000 gives `'medium risk'`
  - anything else gives `'low risk'`
- Print: `Based on income alone, this person is classified as: [risk level]`
- Then also ask for `age`, and combine the two with `and` or `or` so that the
  classification uses both.

In [19]:
income = int(input('Annual income in euros: '))

# Classification based on income alone:
if income < 20000:
    risk = 'high risk'
elif income < 40000:
    risk = 'medium risk'
else:
    risk = 'low risk'

print('Based on income alone, this person is classified as: ' + risk)

# Now also ask for age and combine the two with and / or:
age = int(input('Age: '))

if income < 20000 or age < 21:
    risk = 'high risk'
elif income < 40000 and age < 65:
    risk = 'medium risk'
else:
    risk = 'low risk'

print('Based on income and age, this person is classified as: ' + risk)

Annual income in euros: 25000


Based on income alone, this person is classified as: medium risk


Age: 30


Based on income and age, this person is classified as: medium risk


### 3.4 Reflection

- Is it fair to classify people on income alone? Which information is missing
from this model, and who would be harmed if a real organisation used it?

Your answer: No, income alone is not a fair basis to judge people. It says
nothing about someone's debts, family size, cost of living, health or job
stability, and it ignores why incomes differ - people are paid unequally for
reasons that have nothing to do with how reliable they are. A model like this
would reproduce that bias and harm people with low incomes (for example
students, caregivers and people in low-paid but stable jobs) by denying them
things they could actually pay for. A fairer model would need much more
context, transparent rules, and a way for people to appeal.

---

**BIG INDIVIDUAL ASSIGNMENT · SLIDES 17-19**

## 4. Green or Not?

Everything from the three waves, combined into one program. Individual work. This one goes into your portfolio.

In this assignment you combine everything from today into one program: input,
type conversion, calculation, and a decision. This is a portfolio piece - polish
it at home if you do not finish it in class.

**Data:**

Return short-haul flight: 400 kg CO2 each
Driving: 0.23 kg CO2 per km
Meat-heavy diet: 1400 kg CO2 per year
Sustainable annual budget: 2500 kg CO2

**Classification thresholds:**

Below 2500 kg: under budget
2500 to 4000 kg: slightly over budget
Above 4000 kg: far over budget

**Assignment:**
- Make variables for the four reference values above.
- Ask the user how many return flights they take per year, how many kilometres
  they drive per year, and whether they eat meat daily. Convert each answer to
  the right type.
- Calculate the total annual CO2 and assign it to a variable.
- Use `if` / `elif` / `else` to classify the total against the thresholds.
- Print the type of one of your variables.
- Print a personalised message containing both the total and the classification,
  for example: `You emit 5150 kg CO2 per year. That is far over budget.`

In [20]:
# reference values:
flight_co2 = 400           # kg CO2 per return short-haul flight
car_co2_per_km = 0.23      # kg CO2 per kilometre driven
meat_diet_co2 = 1400       # kg CO2 per year for a meat-heavy diet
sustainable_budget = 2500  # sustainable annual CO2 budget in kg

# questions to the user:
flights = int(input('How many return flights do you take per year? '))
km_driven = int(input('How many kilometres do you drive per year? '))
meat_daily = input('Do you eat meat daily? (yes/no) ').lower() == 'yes'

# calculation:
total = flights * flight_co2 + km_driven * car_co2_per_km
if meat_daily:
    total = total + meat_diet_co2

# classification:
if total < sustainable_budget:      # below 2500 kg
    verdict = 'under budget'
elif total <= 4000:                 # 2500 to 4000 kg
    verdict = 'slightly over budget'
else:                               # above 4000 kg
    verdict = 'far over budget'

# print the result:
print(type(meat_daily))
print('You emit ' + str(round(total)) + ' kg CO2 per year. That is ' + verdict + '.')

How many return flights do you take per year? 2


How many kilometres do you drive per year? 8000


Do you eat meat daily? (yes/no) yes


<class 'bool'>
You emit 4040 kg CO2 per year. That is far over budget.


Your answer: Before a government could use this calculator for a climate tax,
the model would have to be far more complete and transparent: the emission
numbers would have to come from published, regularly updated scientific data;
it would have to cover all major emission sources and not just three; people
would need to be able to check and challenge their result; and there would have
to be exceptions for people who genuinely cannot avoid driving or eating meat.
It would also have to be progressive, so that it does not hit the people who
pollute the least the hardest.